File: Inboxes/Ruby/Cli/inboxes.rb
require 'time'

$messages = []
$next_id = 1

def send_message(from_, subject, body)
  msg = {
    id: $next_id,
    from: from_,
    subject: subject,
    body: body,
    read: false,
    created_at: Time.now.iso8601
  }
  $messages << msg
  $next_id += 1
  msg
end

def list_messages
  $messages
end

def read_message(id)
  msg = $messages.find { |m| m[:id] == id }
  msg[:read] = true if msg
  msg
end

def delete_message(id)
  index = $messages.index { |m| m[:id] == id }
  $messages.delete_at(index) if index
  !index.nil?
end

def main
  loop do
    puts "\n=== Inbox CLI ==="
    puts '1. Send message'
    puts '2. List messages'
    puts '3. Read message'
    puts '4. Delete message'
    puts '5. Exit'
    print 'Choice: '
    choice = gets.to_s.strip

    case choice
    when '1'
      print 'From: '
      from_ = gets.to_s.strip
      print 'Subject: '
      subject = gets.to_s.strip
      print 'Body: '
      body = gets.to_s.strip
      msg = send_message(from_, subject, body)
      puts "Message sent (id=#{msg[:id]})"
    when '2'
      msgs = list_messages
      if msgs.empty?
        puts 'No messages.'
      else
        msgs.each do |m|
          status = m[:read] ? 'R' : 'U'
          puts "[#{m[:id]}] #{status} From: #{m[:from]} | Subject: #{m[:subject]} | #{m[:created_at]}"
        end
      end
    when '3'
      print 'Message ID: '
      id = Integer(gets.to_s.strip, exception: false)
      unless id
        puts 'Invalid ID.'
        next
      end
      msg = read_message(id)
      if msg
        puts "From: #{msg[:from]}"
        puts "Subject: #{msg[:subject]}"
        puts "Date: #{msg[:created_at]}"
        puts "---\n#{msg[:body]}"
      else
        puts 'Message not found.'
      end
    when '4'
      print 'Message ID: '
      id = Integer(gets.to_s.strip, exception: false)
      unless id
        puts 'Invalid ID.'
        next
      end
      if delete_message(id)
        puts 'Message deleted.'
      else
        puts 'Message not found.'
      end
    when '5'
      break
    end
  end
end

main if __FILE__ == $PROGRAM_NAME

File: Inboxes/Python/Cli/inboxes.py
import sys
from datetime import datetime

messages = []
next_id = 1


def send_message(from_, subject, body):
    global next_id
    msg = {
        "id": next_id,
        "from": from_,
        "subject": subject,
        "body": body,
        "read": False,
        "created_at": datetime.now().isoformat(),
    }
    messages.append(msg)
    next_id += 1
    return msg


def list_messages():
    return messages


def read_message(id):
    for m in messages:
        if m["id"] == id:
            m["read"] = True
            return m
    return None


def delete_message(id):
    global messages
    for i, m in enumerate(messages):
        if m["id"] == id:
            messages.pop(i)
            return True
    return False


def main():
    while True:
        print("\n=== Inbox CLI ===")
        print("1. Send message")
        print("2. List messages")
        print("3. Read message")
        print("4. Delete message")
        print("5. Exit")
        choice = input("Choice: ").strip()

        if choice == "1":
            from_ = input("From: ").strip()
            subject = input("Subject: ").strip()
            body = input("Body: ").strip()
            msg = send_message(from_, subject, body)
            print(f"Message sent (id={msg['id']})")
        elif choice == "2":
            msgs = list_messages()
            if not msgs:
                print("No messages.")
            else:
                for m in msgs:
                    status = "✓" if m["read"] else "✗"
                    print(
                        f"[{m['id']}] {status} From: {m['from']} | "
                        f"Subject: {m['subject']} | {m['created_at']}"
                    )
        elif choice == "3":
            try:
                id = int(input("Message ID: ").strip())
            except ValueError:
                print("Invalid ID.")
                continue
            msg = read_message(id)
            if msg:
                print(f"From: {msg['from']}")
                print(f"Subject: {msg['subject']}")
                print(f"Date: {msg['created_at']}")
                print(f"---\n{msg['body']}")
            else:
                print("Message not found.")
        elif choice == "4":
            try:
                id = int(input("Message ID: ").strip())
            except ValueError:
                print("Invalid ID.")
                continue
            if delete_message(id):
                print("Message deleted.")
            else:
                print("Message not found.")
        elif choice == "5":
            break


if __name__ == "__main__":
    main()
